### PR TITLE
Fix frustum cull

### PR DESCRIPTION
**Pull request by tsherif**
Fix frustum culling and create example.

### CODE REVIEW

File: examples/frustum-cull/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<div align="center">
+   <img width="150" heigth="150" src="https://webpack.js.org/assets/icon-square-big.svg" />
+</div>
+
+## Example: Use deck.gl with React
```
Comment (Pessimistress) [inline examples/frustum-cull/README.md]: Name and description of this example

File: examples/frustum-cull/app.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      right: '10px',
+      top: '10px',
+      padding: '10px',
+      background: 'white',
+      border: '1px solid black'
+    };
+  }
+
+  componentDidMount() {
+    const loop = () => {
```
Comment (Pessimistress) [inline examples/frustum-cull/app.js]: Not a big fan of defining large functions inside of functions. Make this a method?

File: examples/frustum-cull/package.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "react-basic",
```
Comment (xintongxia) [inline examples/frustum-cull/package.json]: "react-basic" -> "frustum culling"?

File: examples/frustum-cull/app.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      <DeckGL
+        controller={true}
+        initialViewState={INITIAL_VIEW_STATE}
+        ref={this.deckRef}
+        _animate={true}
+      >
+        <StaticMap
+          key="map"
+          mapStyle="mapbox://styles/mapbox/light-v9"
+          mapboxApiAccessToken="pk.eyJ1IjoidHNoZXJpZiIsImEiOiJjanJ1dm1iNjYwZXM2M3lta2FjZ3dqa3JpIn0.js3Fu08SNC8Ca4VEWeiZ5A"
```
Comment (Pessimistress) [inline examples/frustum-cull/app.js]: Do not include mapbox token

File: examples/frustum-cull/package.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  "version": "0.0.0",
+  "license": "MIT",
+  "scripts": {
+    "start": "webpack-dev-server --progress --hot --open",
+    "start-local": "webpack-dev-server --env.local --progress --hot --open",
+    "build": "webpack -p"
+  },
+  "dependencies": {
+    "deck.gl": "^7.0.0-beta",
+    "luma.gl": "^7.0.0-beta",
```
Comment (Pessimistress) [inline examples/frustum-cull/package.json]: This is unnecessary. You should be able to import the geometry from `@luma.gl/core` which is already a deck dependency.

File: examples/frustum-cull/package.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    "react": "^16.3.0",
+    "react-dom": "^16.3.0"
```
Comment (Pessimistress) [inline examples/frustum-cull/package.json]: Missing `react-map-gl`

File: examples/frustum-cull/app.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
+/* eslint-disable no-continue */
+/* global requestAnimationFrame, cancelAnimationFrame */
+
+import React, {Component} from 'react';
+import {render} from 'react-dom';
+import DeckGL, {SimpleMeshLayer} from 'deck.gl';
+import {StaticMap} from 'react-map-gl';
+import {SphereGeometry} from 'luma.gl';
```
Comment (xintongxia) [inline examples/frustum-cull/app.js]: use submodule instead, `import {SphereGeometry} from '@luma.gl/core';`

File: modules/core/src/utils/math-utils.js [hunk 8]
```diff
@@ -26,33 +27,45 @@ const farCenter = new Vector3();
 const a = new Vector3();
 
 /* eslint-disable max-statements */
-export function getFrustumPlanes({aspect, near, far, fovyRadians, position, direction, up}) {
+
+// Extract frustum planes in common space.
+// Note that common space is left-handed
+// (with y pointing down)
+export function getFrustumPlanes({
```
Comment (xintongxia) [inline modules/core/src/utils/math-utils.js]: Is this API documented somewhere?

File: examples/frustum-cull/app.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+      border: '1px solid black'
+    };
+  }
+
+  componentDidMount() {
+    const loop = () => {
+      if (this.deckRef.current.viewports) {
+        const viewport = this.deckRef.current.viewports[0];
+        const commonPosition = viewport.projectPosition(position);
+        const frustumPlanes = viewport.getFrustumPlanes();
```
Comment (xintongxia) [inline examples/frustum-cull/app.js]: Add some explanations here?

File: examples/frustum-cull/app.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+class Root extends Component {
+  constructor(props) {
+    super(props);
+    this.deckRef = React.createRef();
+    this.animationLoop = null;
+    this.state = {
+      cullStatus: 'IN'
+    };
+
+    this.statusStyle = {
```
Comment (Pessimistress) [inline examples/frustum-cull/app.js]: Make a constant

File: examples/frustum-cull/app.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    this.animationLoop = requestAnimationFrame(this.loop);
+  }
+
+  render() {
+    return (
+      <DeckGL
+        controller={true}
+        initialViewState={INITIAL_VIEW_STATE}
+        ref={this.deckRef}
+        _animate={true}
```
Comment (Pessimistress) [inline examples/frustum-cull/app.js]: Is this necessary?

File: examples/frustum-cull/app.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      padding: '10px',
+      background: 'white',
+      border: '1px solid black'
+    };
+
+    this.loop = this.loop.bind(this);
+  }
+
+  componentDidMount() {
+    this.animationLoop = requestAnimationFrame(this.loop);
```
Comment (Pessimistress) [inline examples/frustum-cull/app.js]: You should be able to subscribe to `onViewStateChange` instead of creating an animation loop

**Comment by tsherif** [comment]
@Pessimistress @xintongxia I think I addressed everything.

File: examples/experimental/frustum-cull/app.js [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    this.deckRef = React.createRef();
+    this.state = {
+      cullStatus: 'IN'
+    };
+
+    this._onViewStateChange = this._onViewStateChange.bind(this);
+  }
+
+  _onViewStateChange() {
+    const viewport = this.deckRef.current.viewports[0];
```
Comment (Pessimistress) [inline examples/experimental/frustum-cull/app.js]: `viewports` have not been updated at this point. You need to manually build the instance:

```js
import {WebMercatorViewport} from '@deck.gl/core';

_onViewStateChange({viewState}) {
  const viewport = new WebMercatorViewport(viewState);
}
```

**Pull request merged**